### DISCUSSION TITLE
How to add text content to a view using a decorator?

### DISCUSSION THREAD

**Discussion by MYDJH** [iTwin/itwinjs-core#7151] [Q&A] [answered]
"Hello, while reading an official example, I tried to add text next to the horizontal line, but it didn't work. Is there any way to achieve this effect?"
 The link to the example:[https://www.itwinjs.org/sandboxes/iTwinPlatform/Curve%20Fractions](url)
This is my code:
```js
curvesData.forEach((curveData) => {

      const curve = curveData.curve;

      const marker = curveData.curvePointMarker;

      const curvePoint = marker.worldLocation;

      const derivative = curveData.derivativeAtPoint;

      decoratorState.setColor(ColorDef.white);
      decoratorState.setFill(false);
      decoratorState.setLineThickness(5);
      decoratorState.setLinePixels(LinePixels.Solid);
      decoratorState.addGeometry(curve);
      decoratorState.addMarker(marker);


      const arrowEnd = curvePoint.plusScaled(derivative, 1);
      decoratorState.setColor(ColorDef.green);
      decoratorState.setFill(false);
      decoratorState.setLineThickness(1);
      console.log(curvePoint,arrowEnd,"arrowEnd")
      decoratorState.addLine(LineSegment3d.create(curvePoint, arrowEnd)); 


      const normalized = derivative.normalize()!;
      const onStem = arrowEnd.plusScaled(normalized, -1.0);
      const perpendicular = normalized.crossProduct(Vector3d.unitZ());

      const triangle = [
        arrowEnd.clone(),
        onStem.plusScaled(perpendicular, 0.5),
        onStem.plusScaled(perpendicular, -0.5),
      ];
      const linestring = LineString3d.create(triangle);
      const loop = Loop.create(linestring.clone());

      decoratorState.setLineThickness(0);
      decoratorState.addGeometry(loop); 

    const verticalLineLength = 10;
    const verticalLineEnd1 = curvePoint.plusScaled(perpendicular, verticalLineLength);
    const verticalLineEnd2 = curvePoint.plusScaled(perpendicular, -verticalLineLength);

    decoratorState.setColor(ColorDef.red); 
    decoratorState.setLineThickness(2);
    decoratorState.setLinePixels(LinePixels.Code3);
    decoratorState.addLine(LineSegment3d.create(verticalLineEnd1, verticalLineEnd2)); 


  const textPosition = curvePoint.plusScaled(perpendicular, -5); 
  const textProps = {
    text: "Text Here",           
    font: 0,                    
    height: 10,                 
    width: 50,                 
    bold: true,                 
    italic: false,              
    underline: false,           
    origin: textPosition,     
    widthFactor: 1,             
  };
  const textString = new TextString(textProps);
  decoratorState.addText(textString);
    });

  }, [decoratorState, fraction, curvesData]);
```
The effect I want to achieve:
![image](https://github.com/user-attachments/assets/9d46b47a-6d7f-4b04-b8f1-ec502ec3ce2f)
Thank you very much for your attention, and I look forward to your suggestions

**Comment by pmconne** [comment]
Your code snippet lacks sufficient context. For example, I don't know what the type of `decoratorState` is.

Your link produces a 404. I assume it contains some of the missing context. Can you fix it please?

Is this code running on the front-end or back-end?

**Reply by MYDJH** [reply]
I'm very sorry for the trouble I caused you，This is the latest link to the example.https://www.itwinjs.org/sandboxes/iTwinPlatform/Curve%20Fractions

**Comment by MYDJH** [comment]
I'm very sorry for the inconvenience I have caused you.
This is the latest link to the example.https://www.itwinjs.org/sandboxes/iTwinPlatform/Curve%20Fractions
Could you please copy and open it in the browser? I'm not sure why clicking directly leads to a 404 error。
You can also find it this way.
https://developer.bentley.com/samples/
![image](https://github.com/user-attachments/assets/aa58eb85-43d3-49ef-a3a7-40a384eabe9d)

"This is the location of the modified code. Replace it with the code above, and you'll be able to see the effect."
![image](https://github.com/user-attachments/assets/62efd819-7f62-400b-a47a-81c2cf1b0a4b)

**Comment by pmconne** [comment, marked as answer]
`TextString` represents text in a geometry stream. Doing anything useful with the text requires access to C and C++ libraries like [freetype](https://freetype.org/), which are not accessible from the front-end.

Browsers are really, really good at rendering text. You can use [canvas decorations](https://www.itwinjs.org/learning/frontend/viewdecorations/#canvas-decorations) or [HTML decorations](https://www.itwinjs.org/learning/frontend/viewdecorations/#html-decorations) to get the browser to efficiently render text into your Viewport.

If you really, truly need to render a `TextString` because the above options do not meet your requirements, and you have access to an iModel on the backend containing the required fonts, you can produce graphics from it using [requestElementGraphics](https://www.itwinjs.org/reference/core-frontend/tiles/tileadmin/requestelementgraphics/).
